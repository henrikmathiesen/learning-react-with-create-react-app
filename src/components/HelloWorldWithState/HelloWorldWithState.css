.HelloWorldWithState {
    border: 2px solid black;
    text-align: center;
    background: #3498db;
    color: black;
    color: #333;
    margin: 20px;
    padding: 20px;
}
