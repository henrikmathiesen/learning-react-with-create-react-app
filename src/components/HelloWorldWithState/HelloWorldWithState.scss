@import '../../_settings';

.HelloWorldWithState {
    border: 2px solid $black;
    text-align: center;
    background: $river;
    color: $almost-black;
    margin: 20px;
    padding: 20px;
}

@import 'foo';
