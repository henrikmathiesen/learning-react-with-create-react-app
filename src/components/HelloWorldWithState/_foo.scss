// dont need to import _settings since parent sass file has it imported

.HelloWorldWithState {
    color:$sun-flower;
}

.testing {
    columns: 100px 4;
}

.testing-too {
    color:pink;
}
