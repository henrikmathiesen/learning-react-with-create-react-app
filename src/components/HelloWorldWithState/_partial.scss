.HelloWorldWithState {
    color: gold;
    font-weight: bold;
}

.test {
    columns: 100px 3;
}
