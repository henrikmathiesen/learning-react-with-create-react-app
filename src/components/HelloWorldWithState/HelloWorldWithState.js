import React, { Component } from 'react';
import './HelloWorldWithState.css';

class HelloWorldWithState extends Component {

    constructor(props) {
        super(props);
        this.state = { greeting: 'Hello there' };

        this.french = this.french.bind(this); // *
    }

    french() {
        this.setState({ greeting: 'Bonjour' });
    }

    render() {
        return (
            <div className="HelloWorldWithState">
                <div>
                    {this.state.greeting} {this.props.name} World!
                </div>
                <div>
                    <button onClick={this.french}>Frenchify!</button>
                </div>
            </div>
        );
    }
}

export default HelloWorldWithState;


// Import the default export from 'react' in node_modules as React.
// Also import named export Component from 'react'.
// We use a class here, because this component focuses on more than just rendering content.
// We need to define a render function in our class for rendering JSX.
// This component can also get props passed to it, but since this is a class we get those in the constructor,
// that means that we have to use this.props.name instead of props.name

// OBS
// A simple constructor like this is not needed, it seems that it is done automaticly, and in fact the linter complains about this as Useless constructor
// constructor(props) {
//  super(props);
// }

// HOWEVER
// If we use state, we set that up in the constructor, lite above

// Changing state
// Any state changes NEED to happen via the this.setState function

// * This tells Javascript "Hey, any time you see a reference to this inside of the frenchify function, I want you specifically to refer to ME."
