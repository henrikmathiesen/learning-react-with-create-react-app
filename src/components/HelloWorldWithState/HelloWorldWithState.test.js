import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import HelloWorldWithState from './HelloWorldWithState';

// Can actually reference the component like this HelloWorldWithState (without quotes) but this creates noise in test command prompt
describe('HelloWorldWithState', () => {

    beforeAll(() => {
        // If console gives error, test fails
        console.error = error => {
            throw new Error(error);
        };
    });

    // Set up a shallow rendered component with enzyme api
    const name = 'Person';
    const mockRemoveGreeting = jest.fn();
    const component = shallow(
        <HelloWorldWithState name={name} removeGreeting={mockRemoveGreeting} />
    );

    // 1) Write a generic snapshot test (these are not shallow, if we look at the snapshot of a component that has children (HelloWorldList for example) we can see that)

    it('Renders and matches our snapshot', () => {
        const component = renderer.create(
            <HelloWorldWithState name="Person" />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    // 2) Write some sub-component-specific tests (skip since we dont have any sub components)

    // 3) Check content

    it('Contains the supplied name', () => {
        expect(component.text()).toContain(name);
    });

    // 4) Check interactions

    it('Modifies the greeting when frenchify button is clicked', () => {
        component.find('button.ts-french').simulate('click');
        expect(component.text()).toContain('Bonjour');
    });

    it('Calls the passed in removeGreeting function when remove button is clicked', () => {
        component.find('button.ts-remove').simulate('click');
        expect(mockRemoveGreeting).toBeCalled();
    });
});


// We installed some new packages for this (npm throwed error, needed to do npm cache clean / run bash as admin)

// react-test-renderer
// - Render components to pure JavaScript objects, without depending on the DOM or a native mobile environment
// - We can create a JSON from it and make a snapshot for snapshot testing

// enzyme
// - More flexibility, like shallow rendering (test components in isolation without children), although react-addons-test-utils also has this
// - Also mimicking jQuery's API for DOM manipulation and traversal
// - Seems more easy to use than react-addons-test-utils
// - "Using Enzyme gives you the same functionality as ReactTestUtils but with much cleaner, simpler commands"
// - https://medium.com/@malinnaleach/testing-react-a-beginners-guide-755a8828c222#.1bcemeuvl

// react-addons-test-utils
// - enzyme requires this package
// - It has its own test api, see HelloWorld.test.js

// snapshot testing: https://facebook.github.io/jest/blog/2016/07/27/jest-14.html
// press u in test runner to update snap shot if change was intentional
