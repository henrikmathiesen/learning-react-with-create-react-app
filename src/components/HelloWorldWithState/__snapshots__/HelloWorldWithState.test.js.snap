exports[`HelloWorldWithState Renders and matches our snapshot 1`] = `
<div
  className="HelloWorldWithState">
  <div>
    Hello there
     
    Person
     World!
  </div>
  <div>
    <button
      className="ts-french"
      onClick={[Function]}>
      Frenchify!
    </button>
    <button
      className="ts-remove"
      onClick={[Function]}>
      Remove
    </button>
  </div>
</div>
`;
